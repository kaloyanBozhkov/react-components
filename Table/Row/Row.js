import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

// import components
import Icon from 'UI/Icon/Icon'

import styles from './styles.module.scss'

// import helpers
import { dateDisplay } from '~/helper/date'

const Row = ({ rowCells, pageName, uniqueRowId, withActions = true }) => {
  const classes = [styles.tableRow, withActions ? styles.withActions : '', styles[pageName] || '']
    .join(' ')
    .trim()

  return (
    <div className={classes}>
      {rowCells.map(({ cellValue, cellType, centered = false, onClick = null }, index) => {
        const key = `${uniqueRowId}-${index}`
        const centeredCellContents = centered ? styles.centeredCellContents : undefined
        const onClickFunc = onClick ? () => onClick({ cellValue, rowIndex: index }) : undefined

        switch (cellType) {
          case 'text':
            return (
              <div
                key={key}
                className={centeredCellContents}
                onClick={onClickFunc}
                role={onClickFunc ? 'button' : undefined}
              >
                <p>{cellValue}</p>
              </div>
            )
          case 'icon':
            return (
              <div
                icon={cellValue}
                key={key}
                onClick={onClickFunc}
                role={onClickFunc ? 'button' : undefined}
              >
                <Icon icon={cellValue} className={centeredCellContents} />
              </div>
            )
          case 'link':
            if (
              typeof cellValue === 'object' &&
              cellValue.hasOwnProperty('path') &&
              cellValue.hasOwnProperty('label')
            ) {
              return (
                <div key={key} className={styles.urlDiv} onClick={onClickFunc}>
                  <Link className={centeredCellContents} to={cellValue.path}>
                    {cellValue.label}
                  </Link>
                </div>
              )
            }

            // or return a UI friendly msg?
            return null
          case 'date': {
            const parseDateStringOrDate = new Date(cellValue)

            // check if invalid date obj
            if (Number.isNaN(parseDateStringOrDate.getTime())) {
              // or return a UI friendly msg?
              return null
            }

            return (
              <div
                role={onClickFunc ? 'button' : undefined}
                key={key}
                className={[styles.dateDiv, centeredCellContents].join(' ').trim()}
                onClick={onClickFunc}
              >
                <p>
                  <b>{dateDisplay(parseDateStringOrDate).withShortMonths().format()}</b>
                </p>
                <p>{parseDateStringOrDate.toLocaleTimeString()}</p>
              </div>
            )
          }
          default:
            return (
              <div
                key={key}
                className={centeredCellContents}
                onClick={onClickFunc}
                role={onClickFunc ? 'button' : undefined}
              >
                <p>-</p>
              </div>
            )
        }
      })}
    </div>
  )
}

Row.propTypes = {
  rowCells: PropTypes.arrayOf(
    PropTypes.shape({
      cellValue: PropTypes.oneOfType([
        //label/text
        PropTypes.string,
        //timestamp
        PropTypes.number,
        //link obj
        PropTypes.shape({
          path: PropTypes.string,
          label: PropTypes.string,
        }),
        //other proptypes based on added cell types for row generation
      ]).isRequired,
      cellType: PropTypes.oneOf(['text', 'icon', 'link', 'date', 'img']),
      centered: PropTypes.bool,
      onClick: PropTypes.func,
    })
  ).isRequired,
  pageName: PropTypes.string.isRequired,
  uniqueRowId: PropTypes.string,
  withActions: PropTypes.bool,
}

export default Row
