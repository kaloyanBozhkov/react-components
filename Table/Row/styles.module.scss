@import 'scss/abstracts/_variables';
.tableRow {
  display: grid;
  grid-gap: 1px;

  background-color: #eaecef;

  &:first-of-type {
    padding: 1px;
  }

  &:not(:first-of-type) {
    padding: 0px 1px 1px 1px;
  }

  &:hover {
    > div {
      background-color: #f5f5f5;
    }
  }
  > div {
    display: flex;
    align-items: center;

    font-size: 1.2rem;
    transition: 200ms;
    background-color: white;

    padding: $medium-padding * 1.5;

    &.centeredCellContents {
      justify-content: center;
    }

    &.urlDiv {
      a {
        color: $secondary-color;
        text-decoration: none;
        &:hover {
          color: darken($secondary-color, 15%);
        }
      }
    }

    &.dateDiv {
      flex-direction: column;
      align-items: flex-start;
      &.centeredCellContents {
        align-items: center;
      }
      p {
        &:first-of-type {
          font-weight: bold;
        }
        &:last-of-type {
          font-weight: normal;
        }
      }
    }

    &[icon] {
      //icons and images center here!!
      justify-content: center;
    }
  }

  &.withActions {
    > div {
      &[icon='edit'],
      &[icon='trash-alt'],
      &[icon='trash'] {
        cursor: pointer;
        &:hover {
          svg {
            transform: scale(0.85);
          }
        }
      }

      &[icon='edit'] {
        color: $primary-purple;
      }

      &[icon='check'] {
        color: $success-color;
      }

      &:nth-child(2),
      &:nth-child(3) {
        min-width: 5rem;
      }
    }
  }

  //CUSTOM STYLINGS BASED ON CONTENT HERE
  // &.games {
  //   grid-template-columns: repeat(2, 8.5rem) 1fr 0.67fr repeat(3, 8.5rem);
  //   grid-template-rows: minmax(7.5rem, auto);
  // }
  // &.gameFilters {
  //   grid-template-columns: repeat(2, 1fr) repeat(2, 8.5rem);
  //   grid-template-rows: auto;
  // }
  // &.components {
  //   grid-template-columns: repeat(2, 1fr) repeat(3, 8.5rem);
  //   grid-template-rows: auto;
  // }
  
  &.accountSummary {
    grid-template-columns: repeat(4, 1fr);
  }
}
