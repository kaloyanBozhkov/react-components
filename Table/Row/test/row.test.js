import React from 'react'
import { shallow } from 'enzyme'

import Row from '../Row'

describe('testing Row component', () => {
  it('should render Row', () => {
    const rowCells = [
      {
        cellValue: 'hello world',
      },
    ]

    const wrapper = shallow(
      <Row
        rowCells={rowCells}
        pageName="somePageModifier"
        uniqueRowId="uniqueRowId"
        uniqueRowId="uniqueRowId"
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render Row without withAction class', () => {
    const rowCells = [
      {
        cellValue: 'hello world',
      },
    ]

    const wrapper = shallow(
      <Row
        rowCells={rowCells}
        withActions={false}
        pageName="somePageModifier"
        uniqueRowId="uniqueRowId"
      />
    )

    expect(wrapper.hasClass('withActions')).toEqual(false)
  })

  it('should render Row without pageName specific class', () => {
    const rowCells = [
      {
        cellValue: 'hello world',
      },
    ]

    const wrapper = shallow(
      <Row rowCells={rowCells} pageName="somePageModifier" uniqueRowId="uniqueRowId" />
    )

    expect(wrapper.hasClass('somePageModifier')).toEqual(true)
  })

  it('should render Row with default cell value', () => {
    const rowCells = [
      {
        cellValue: 'hello world',
      },
    ]

    const wrapper = shallow(
      <Row rowCells={rowCells} pageName="somePageModifier" uniqueRowId="uniqueRowId" />
    )

    //test default label when misconfigured cell
    expect(wrapper.childAt(0).text()).toEqual('-')
  })

  describe('testing for text type cells', () => {
    it('should render Row with text cell type', () => {
      const rowCells = [
        {
          cellValue: 'hello world',
          cellType: 'text',
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )

      //test default label when configured cell
      expect(wrapper.childAt(0).text()).toEqual('hello world')
    })

    it('should render Row with text cell type wiht optional class', () => {
      const rowCells = [
        {
          cellValue: 'hello world',
          cellType: 'text',
          centered: true,
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )
      const rowCell = wrapper.childAt(0)

      //test default label when configured cell
      expect(rowCell.hasClass('centeredCellContents')).toEqual(true)
    })

    it('should render Ropw with text cell type wiht optional click function', () => {
      const mockFunc = jest.fn(() => 'works')
      const rowCells = [
        {
          cellValue: 'hello world',
          cellType: 'text',
          onClick: mockFunc,
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )
      const rowCell = wrapper.childAt(0)
      rowCell.simulate('click')

      //test default label when configured cell
      expect(mockFunc).toHaveBeenCalled()
    })
  })

  describe('tests for Icon cell type', () => {
    //do tests for click, opt classes? They are the same as text, just contents change. If we should, put them here! Otherwise, remove the describe block?

    it('should render Row with Icon cell type', () => {
      const rowCells = [
        {
          cellValue: 'check',
          cellType: 'icon',
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )

      expect(wrapper.find('div[icon="check"]')).toHaveLength(1)
      expect(wrapper.text()).toEqual('<Icon />')
    })
  })

  describe('testing Row rendering link type cells', () => {
    it('should render Row with link cell type', () => {
      const rowCells = [
        {
          cellValue: {
            path: 'someRoute',
            label: 'redirectMe',
          },
          cellType: 'link',
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )
      const rowCell = wrapper.childAt(0)

      expect(rowCell.hasClass('urlDiv')).toEqual(true)
    })

    it('should handle wrong obj format for link', () => {
      const rowCells = [
        {
          cellValue: {
            wrong: 'redirectMe',
          },
          cellType: 'link',
        },
      ]

      //@TO DO: handle wrong prop types?! https://medium.com/shark-bytes/type-checking-with-prop-types-in-jest-e0cd0dc92d5
      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )

      //dont render Link if
      expect(wrapper.children()).toHaveLength(0)
    })
  })

  describe('test Row rendering with date cells', () => {
    it('should render Row with date cell type', () => {
      const rowCells = [
        {
          cellValue: new Date('1990'),
          cellType: 'date',
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )
      const rowCell = wrapper.childAt(0)

      expect(rowCell.hasClass('dateDiv')).toEqual(true)
      expect(rowCell.children()).toHaveLength(2)
      expect(rowCell.childAt(0).childAt(0).text()).toEqual('1st Jan 1990')
    })

    it('should render Row with date cell type and string date', () => {
      const rowCells = [
        {
          cellValue: '1/1/1990',
          cellType: 'date',
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )
      const rowCell = wrapper.childAt(0)

      expect(rowCell.hasClass('dateDiv')).toEqual(true)
      expect(rowCell.children()).toHaveLength(2)
      expect(rowCell.childAt(0).childAt(0).text()).toEqual('1st Jan 1990')
    })

    it('should render Row with date cell type and timestamp date', () => {
      const rowCells = [
        {
          cellValue: 631152000000,
          cellType: 'date',
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )
      const rowCell = wrapper.childAt(0)

      expect(rowCell.hasClass('dateDiv')).toEqual(true)
      expect(rowCell.children()).toHaveLength(2)
      expect(rowCell.childAt(0).childAt(0).text()).toEqual('1st Jan 1990')
    })

    it('should render Rowwith date cell type and fail due to invalid date', () => {
      const rowCells = [
        {
          cellValue: 'asd',
          cellType: 'date',
        },
      ]

      const wrapper = shallow(
        <Row pageName="somePageModifier" uniqueRowId="uniqueRowId" rowCells={rowCells} />
      )

      //test that the rendered row's cell has no contents, since return null when date is invalid
      expect(wrapper.children()).toHaveLength(0)
    })
  })
})
