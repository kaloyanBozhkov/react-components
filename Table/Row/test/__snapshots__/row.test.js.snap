// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Row component should render Row 1`] = `
<div
  className="tableRow withActions somePageModifier"
>
  <div
    key="uniqueRowId-0"
  >
    <p>
      -
    </p>
  </div>
</div>
`;
